tools/discovery/discovery.go

```diff
@@ -0,0 +1,131 @@
+package discovery
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+
+	"github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
```

Comment:
I think the library packge(s) should be providing the tools, and should not be bound to a particular command line library.  Following the Python example would result in the following Go packages:

* `host_based_image_names`, exposing [`Parse(s string) map[string]string`][1].
* A ref-engine registry package that allows consumers to supply their own ref-engine handlers for new protocols, or clobber the stock handlers for existing protocols.
* An OCI Index Template implementation package (optionally sharing the same namespace as the ref-engine registry package).
* A ref-engine-discovery package, which consumes the host-based-image-name parser and the ref-engine registry.

The command-line bindings in `tools/cmd` would call the CLI-agnostic ref-engine-discovery API using vanilla Go, instead of passing through a `*cli.Context`.

I'd recommend anything that could be backed by network access take a `Context` from `golang.org/x/net/context` (like [this][2]) so callers could gracefully abort.  But that's not a CLI context.

[1]: https://github.com/xiekeyang/oci-discovery/commit/ce7083d030142ac71884cb4b1a95bdadab4eace9#commitcomment-24312383
[2]: https://github.com/opencontainers/image-tools/pull/40/files#diff-a372a96ac413817b9f8f567e310b60bcR29